tpl/template_funcs.go

```diff
@@ -1392,6 +1392,47 @@ func replace(a, b, c interface{}) (string, error) {
 	return strings.Replace(aStr, bStr, cStr, -1), nil
 }
 
+// partialCache represents a cache of partials protected by a mutex.
+type partialCache struct {
+	mu sync.RWMutex
+	p  map[string]template.HTML
+}
+
+// Get retrieves partial output from the cache based upon the partial name.
+// If the partial is not found in the cache, the partial is rendered and added
+// to the cache.
+func (c *partialCache) Get(key, name string, context interface{}) (p template.HTML) {
+	var ok bool
+
+	if p, ok = c.get(key); !ok {
+		p = c.set(key, name, context)
+	}
+
+	return p
+}
+
+func (c *partialCache) get(key string) (p template.HTML, ok bool) {
+	c.mu.RLock()
+	p, ok = c.p[key]
+	c.mu.RUnlock()
+	return
+}
+
+func (c *partialCache) set(key, name string, context interface{}) template.HTML {
```

Comment:
There is still a window between get/set where the partial could have been created by others. That is why I commented on the get/set construction. You can fix it by adding an additional map lookup inside the write lock, but then the set/get method names will be slightly misleading. As such I prefer reading the whole thing as one big code block. You would still need a double-check inside the write lock, but mistakes will be easier to spot.

Note that in normal cache situations (like a regexp cache), I may leave out the double-check because it doesn't matter if it gets invoked 1 or 2 times -- but in this case it has potential to get unexpected results (different partials).